---
title: "HSC simulation results"
output:
  word_document:
      toc: true
---

```{r setup, echo = FALSE, include = FALSE}
knitr::opts_chunk$set(echo = FALSE)
knitr::opts_chunk$set(dev = "png", dev.args = list(type = "cairo-png"))
```



```{r data, include = FALSE}
suppressPackageStartupMessages({
    library(dplyr)
    library(ggplot2)
    library(glue)
    library(ggpubr)
    library(purrr)
})

date = "2021_05_03"
results = vroom::vroom(glue("HSC_sim_results_{date}.csv"))
```
## PACER Simulation Parameters 

We assume that the accumulation of passenger mutations is described by a Poisson birth-death 
stochastic process. As the birth and death rates scale with the number of HSCs, we assume a 
linear birth-death process. 

We assume that the birth rate for a given hematopoietic stem cell (HSC) $i$ at time $t$ with fitness $s_i(t)$) is 
$\lambda_i(t) \sim Poisson(\omega * X_i(t) * (1 + s_i(t)) * dt)$, where $dt$ represents the amount of time in years, and $\omega$ represents the number of stem cell divisions per year. We assume that the death rate can be described as $$\psi_i(t) \sim Poisson(\omega * X_i(t) * (1 - s_i(t)) * dt)$$. The death 
rate is the rate at which an HSC divides into two differentiated cells, and the birth rate is the rate at which
an HSC divides into two HSCs. We don't consider asymmetric HSC differentiation as this would not change
the clone size. The HSC clone cell count is defined as $X_i(t) = \sum_{l \leq t} \lambda_i(l) - \psi_i(l)$, 
and the HSC clone size (a fraction of the total cell population) is $VAF_i(t) = \frac{X_i(t)}{\sum_j X_j(t)}$. 

We start with 500 HSC clones, each with 200 identical cells in each clone $X_i(t = 0) = 200$. Each cell divides once every three years (\omega = 1/3), 
and each clone with an initial $s_i(t = 0) = 0$. At each iteration, we also center the $s_i(t)$ such that 
$\bar{s_i(t)} = 0$. This means that there are 100,000 total HSCs at the start of the simulation. 

For each clone, we set the passenger mutation rate:
 
  1. $\mu_p$, the passenger accumulation rate, $A_i(t) \sim Poisson(X_i(t) * \mu_p * dt)$  

Where $A_i(t)$ is the number of passengers accumulated in a given clone through time $t$. We
set $\mu_p = 0.006$, which is the passenger mutation rate of a diploid genome for a single HSC per year. 
This implies a mutation rate of 6 passengers per year for a clone with 1000 cells, and a mutation rate
of 600 passengers per year across the entire population of 100,000 HSCs. We will later consider the effects
of an insensitive sequencing assay that captures a small fraction of the passengers. 

We assign a single driver to one of the HSC clones, which is randomly selected among the HSC clones.
The time of acquisition is uniformly drawn from each cell division after 10 years, such that are driver is equally likely to be acquired at either 10 years or 78 years. We simulate the HSC population across a lifetime of 90 years. We refer to the time of driver acquision as $T_d$.

We assume that each HSC clone can at most acquire a single driver, which represents a similar HSC
population to the TOPMed CHIP driver carriers.

If an HSC clone $i$ acquires a driver at time $t$, we set $s_i(t) = Beta(4, 16)$. A $Beta(4, 16)$ random 
variable is bounded between 0 and 1 and has an expectation of 0.20. An HSC with $_i(t) = 0.20$ will
self-renew 60\% of the time, and terminally differentiate 40\% of the time. 

For a given HSC population, we simulate 90 years, and track the accumulation of passengers and drivers. 
To incorporate the censoring from using 38x sequencing coverage, we simulate whether a given passenger would
be observed at 38x coverage by sampling the number of alt-reads from $R \sim Binomial(38, VAF_i(t))$ and comparing $R \geq 2$, since two reads are 
required by our variant calling process. We refer to $P(R \geq 2 | VAF = vaf) = P(Binomial(38, vaf) \geq 2)$ We refer to the passengers that would be detected at 38x coverage
as the censored founding passengers, $AC_i(t)$, where $t = T_d$ .  

We ran the simulation 10,000 times, where at most a single HSC clone acquires a driver mutation. We 
then compared $AC_i(T_d)$ to the fitness of the clone at the end of each simulation.

We make a few observations:

For a homogenous birth-death process $Y(t)$, it has previously been established that 
$E(Y(t)) = Y(0)e^{(\lambda - \psi) * t}$, where $Y(0)$ is the initial population size. In our case, once a driver has been acquired,
$\lambda - \psi = \omega * Y(0) * 2s$ . Therefore we can construct an upper bound for the expectation of clone size over 90 years:

$E(X_i(t = 90)) \leq X_i(0)e^{\omega * X_i(0) * 2s * 90}$ and we can approximate

$E(X_i(t)) \simeq \sum_{z=10}^{t} P(T_d <= z) * E(X_i(t; s > 0)) + P(T_d > z) * E(X_i(t; s = 0)) = \sum_{z=10}^{t} \frac{z - 10}{t - 10} * X_i(0)e^{\omega * X(0) * 2s * (t - z)} + \frac{t - z}{t - 10} * X_i(0)$

when $t \geq 10$. 

For $t \geq 10$, then 


$E(AC_i(t)) = \int_{ac_i(t) = 0}^{\infty} \int_{vaf=0}^{0.5} ac_i(t) * P(AC_i(t) = ac_i(t) | VAF_i(t) = vaf_i(t)) * P(VAF_i(T) = vaf_i(t)) dvaf_i(t) dac_i(t)$


$E(AC_i(t) | VAF_i(t) = vaf_i(t)) \simeq E(X_i(t)) * \mu_p * dt * P(R \geq 2 | VAF = vaf_i(t))$

We can approximate $P(VAF = vaf_i(t))$ by applying beta regression to the clone sizes from our simulation. 

This results in $P(VAF = vaf_i(t)) = P(Beta(\alpha, \beta) = vaf_i(t))$ when $T_d < t$, i.e, $i$ has already acquired a driver. When $T_d > t$, we have
$P(VAF = vaf_i(t)) = P(Beta(\phi * \sigma(-2.7), \phi - \phi * \sigma(-2.7)) = vaf_i(t))$,

where $\alpha = \phi * \sigma(-2.7 + 5.1s + 0.023 * (t - T_d))$ and $\beta = \phi - \alpha$ .  $\hat{\phi} = 4.7$
as estimated from the beta regression. $\sigma$ is defined as the sigmoid function. 

Therefore,

$E(AC_i(t)) \simeq \mu_p * \int_{vaf=0}^{0.5} (\sum_{z=10}^{t} \frac{z - 10}{t - 10} * X_i(0)e^{\omega * X_i(0) * 2s * (t - z)} + \frac{t - z}{t - 10} * X_i(0)) * P(R \geq 2 | VAF = vaf_i(t)) * \frac{\Gamma(\alpha)\Gamma(\beta)}{\Gamma(\alpha + \beta)} vaf_i(t)^{\alpha - 1} (1 - vaf_i(t))^{\beta - 1} dvaf_i(t) dt$

if the driver is acquired relatively early in life after 10 years old. 

```{r, include = FALSE}
vaf_model = results %>%
    select(lastVAF, prevVAF, fitness, driver_age) %>%
    tidyr::gather(key = time, value = vaf, 1:2) %>%
    mutate(
        time = ifelse(time == "lastVAF", 90, 80),
        driver_age = ifelse(time == 80, driver_age - 10, driver_age),
        vaf = pmin(0.999, pmax(vaf, 0.001))
    ) %>%
    betareg::betareg(vaf ~ fitness + driver_age, link = "logit", data = .)

expectation = function(
    s = 0.001,
    t = 90,
    td = 30,
    phi = 3.755,
    division_rate = 0.33,
    init_clone_size = 200,
    mutation_rate = 0.006,
    scaling_constant = 1e6,
    n_sims = 1e4
) {
    intercept = -3.24
    fitness_slope = 6.19
    driver_age_slope = 0.012

    #alpha = phi * exp(-5.3 + 5 * s + 0.024 * (t - td))
    alpha = phi * plogis(intercept + fitness_slope * s + driver_age_slope * (t - td))
    beta = phi - alpha

    beta_sims = rbeta(n = n_sims, shape1 = alpha, shape2 = beta)
    censor = pbinom(2, size = 38, prob = beta_sims, lower.tail = FALSE)

    z = 10:t
    uncensored_count = sum(
        ((z - 10) / (t - 10)) * init_clone_size * exp(division_rate * init_clone_size * 2 * s * (t - z)) + 
        ((t - z) / (t - 10)) * init_clone_size
    )
    return(mutation_rate * uncensored_count * censor * t / scaling_constant)
}

analytical_test = results %>%
    slice(1:500) %>%
    mutate(
        analytical_prediction = pmap_dbl(
            list(s = fitness, td = driver_birth),
            function(s, td) mean(expectation(s = s / 100, t = 90, td = td))
        )
    )
```

##  PACER Simulation Results

### Supplementary Figure S6

```{r plot1, echo = FALSE, message = FALSE, warning = FALSE}
p = ggscatter(results %>% mutate(passengers_before_driver_censored = log1p(passengers_before_driver_censored)), 
        x = "passengers_before_driver_censored", y = "fitness",
        color = "black", size = 3, # Points color, shape and size
        alpha = 0.05,
        add = "reg.line",  # Add regressin line
        add.params = list(color = "blue", fill = "lightgray"), # Customize reg. line
        conf.int = TRUE, # Add confidence interval
        cor.coef = TRUE, # Add correlation coefficient. see ?stat_cor
        cor.coeff.args = list(method = "spearman", label.x = 10, label.y = 0.10, label.sep = "\n")
    ) +
    cowplot::theme_cowplot(font_size = 14) +
    labs(x = "log(1 + Founding censored passengers)", y = "Driver fitness")
p
```

We observed a modest concordance between the founding censored passengers with the fitness
of each HSC at the end of the simulation (spearman \rho = 0.09, pvalue < 2.2e-16, Supplementary Figure S6).
This suggests that the censored founding passengers are proportional to the fitness of the driver mutations. 
Stochastic drift of the HSC clone sizes contributes substantial residual variance. 

### Supplementary Figure S7

We observe no concordance between the uncensored founding passengers with fitness (Supplementary Figure S7). 

```{r plot2, echo = FALSE, message = FALSE, warning = FALSE}
p = ggscatter(results %>% mutate(passengers_before_driver = log1p(passengers_before_driver)), 
        x = "passengers_before_driver", y = "fitness",
        color = "black", size = 3, # Points color, shape and size
        alpha = 0.2,
        add = "reg.line",  # Add regressin line
        add.params = list(color = "blue", fill = "lightgray"), # Customize reg. line
        conf.int = TRUE, # Add confidence interval
        cor.coef = TRUE, # Add correlation coefficient. see ?stat_cor
        cor.coeff.args = list(method = "spearman", label.x = 10, label.y = 0.10, label.sep = "\n")
    ) +
    cowplot::theme_cowplot(font_size = 14) +
    labs(x = "log(1 + Founding passengers)", y = "Driver fitness")
p
```

```{r, include = FALSE}
results %>%
    MASS::glm.nb(passengers_before_driver_censored ~ driver_birth + I(fitness * driver_age), data = .) %>%
    broom::tidy(.)
```

### Competing clone simulation demonstrates rs2887399 advantage for DNMT3A

We simulate two competing clones, where in a given simulation, exactly two drivers are acquired.
These two simulated drivers represent DNMT3A and TET2. Here, the fitness of the DNMT3A clone
$s_{i, dnmt} ~ Beta(10, 25)$, which means that $E(s_{i, dnmt}) = 0.286$. 
We will run this simulating in two settings to simulate different rs2883799 genotypes - 

 1. rs2883799 = T/T, where TET2 will have 20% less fitness on average than DNMT3A  
 2. rs2883799 = G/G, where TET2 will have 20% more fitness on average than DNMT3A  

We run the simulation such that the two clone are competing - i.e., the two drivers
are assigned to separate HSC clones. We then track the effects of the different
fitness settings across the simulated HSC lifetime of 90 years. 

```{r competing_clone, echo = FALSE, message = FALSE, warning = FALSE}
competing_clone_date = "2021_10_27"
tet2_strong = vroom::vroom(glue("HSC_sim_results_two_drivers_tet2_strong_{competing_clone_date}.csv"))
tet2_weak   = vroom::vroom(glue("HSC_sim_results_two_drivers_tet2_weak_{competing_clone_date}.csv"))

clone_sim = bind_rows(
    tet2_strong %>% mutate(rs2887399 = "G/G"),
    tet2_weak %>% mutate(rs2887399 = "T/T")
) 

ggplot(data = clone_sim, aes(y = rs2887399, x = log10(passengers_before_driver_censored + 1), fill = driver_label)) +
    ggridges::geom_density_ridges(alpha = 0.7) +
    labs(x = "log10(1 + Founding censored passengers)", fill = "Driver") + 
    cowplot::theme_cowplot(font_size = 12)
```
